README.pod
lib/App/derived.pm

Changes
Revision history for Perl extension App::derived

0.09    Web Feb 20 1:09:31 2013
        - use relative "bash" path instead of absolute path.
          some *BSD system install bash to /usr/bin 

0.08    Tue Feb 19 9:51:33 2013
        - lower compatibility
          If no plugins are loaded, Memcached plugins will be loaded by default

0.07    Tue Feb 19 00:58:26 2013
        - requires LWP
        - supports prefix in GrowthForecast

0.06    Tue Feb 19 00:34:03 2013
        - supports match_key in GrowthForecast
        - generates unique tagname for plugins

0.05    Mon Feb 18 17:24;02 2013
        - [IMPORTANT] imcompatible change
          Changed command options of derived
          Added Pluggable system. New plugins are Memcached, GrowthForecast and Dumper.

0.04    Thu Feb 14 11:28:37 2013
        - randomize timing of starting command worker
        - stopped use preforking server
        - can get last values with ':latest' suffix

0.03    Tue Feb 12 23:32:27 2013
        - initialize persec with 0E0.
          IF there is no previous data to calculate rate, the server returns "0E0"

0.02    Tue Feb 12 11:31:57 2013
        - requires '/bin/bash'

0.01    Thu Feb  7 09:52:05 2013
        - original version


Build.PL
use strict;
use warnings;
use Module::Build;
use Carp;

requires_external_bin('/bin/bash');

my $build = Module::Build->new(
    license              => 'perl',
    recursive_test_files => 1,
    dynamic_config       => 0,

    build_requires       => {
        'Test::More' => '0.98',
        'Test::Requires' => 0,
    },
    configure_requires   => { 'Module::Build' => '0.38' },
    requires             => {
        perl => '5.008005',
        'Getopt::Long'                      => '0',
        'Pod::Usage'                        => '0',
        'File::Temp'                        => '0',
        'File::Copy'                        => '0',
        'IO::Socket::INET'                  => '0',
        'POSIX'                             => '0',
        'Proclet'                           => '0.23',
        'JSON'                              => '2.53',
        'Log::Minimal'                      => '0.14',
        'Module::Load'                      => '0.24',
        'Class::Accessor::Lite'             => '0.05',
        'LWP::UserAgent'                    => '6.02',
    },

    no_index    => { 'directory' => [ 'inc' ] },
    name        => 'App-derived',
    module_name => 'App::derived',

    script_files => ['bin/derived'],

    test_files => (-d '.git' || $ENV{RELEASE_TESTING}) ? 't/ xt/' : 't/',

    create_readme  => 1,
    create_license => 1,
);
$build->create_build_script();

# Module::Install::Can

sub can_run {
    my $cmd = shift;
    require ExtUtils::MakeMaker;
    if ( $^O eq 'cygwin' ) {
        
        # MM->maybe_command is fixed in 6.51_01 for Cygwin.
        ExtUtils::MakeMaker->import(6.52);
    }
    
    my $_cmd = $cmd;
    return $_cmd if ( -x $_cmd or $_cmd = MM->maybe_command($_cmd) );
    
    for my $dir ( ( split /$Config::Config{path_sep}/x, $ENV{PATH} ), q{.} )
        {
            next if $dir eq q{};
            my $abs = File::Spec->catfile( $dir, $cmd );
            return $abs if ( -x $abs or $abs = MM->maybe_command($abs) );
        }
    
    return;
} ## end sub can_run

# Module::Install::External

sub requires_external_bin {
    my ( $bin, $version ) = @_;
    if ($version) {
        croak 'requires_external_bin does not support versions yet';
        
    }

    # Locate the bin
    print "Locating required external dependency bin: $bin...";
    my $found_bin = can_run($bin);
    if ($found_bin) {
        print " found at $found_bin.\n";
    } else {
        print " missing.\n";
        print "Unresolvable missing external dependency.\n";
        print "Please install '$bin' seperately and try again.\n";
        print {*STDERR}
            "NA: Unable to build distribution on this platform.\n";
        exit 0;
    }
    
    return 1;
} ## end sub requires_external_bin


lib/App/derived/Plugin.pm
package App::derived::Plugin;

use strict;
use warnings;
use Class::Accessor::Lite (
    new => 1,
    ro => [qw/_services _proclet/]
);
use JSON qw//;

my $_JSON = JSON->new()
    ->utf8(1)
    ->shrink(1)
    ->space_before(0)
    ->space_after(0)
    ->indent(0);

sub json {
    $_JSON;
}

sub exists_service {
    my $self = shift;
    my $key = shift;
    exists $self->_services->{$key};
}

sub service_stats {
    my $self = shift;
    my $key = shift;
    return if ! exists $self->_services->{$key};
    my $service = $self->_services->{$key};
    open my $fh, '<', $service->{file} or return;
    my $val = do { local $/; <$fh> };
    $self->json->decode($val);
}

sub service_keys {
    my $self = shift;
    keys %{$self->_services};
}

sub add_worker {
    my $self = shift;
    my ($tag, $code, $workers) = @_;
    $workers ||= 1;
    $self->_proclet->service(
        code => $code,
        tag => $tag,
        worker => $workers
    );
}

sub init {
    die 'abstract method';
}


1;

__END__

=encoding utf8

=head1 NAME

App::derived::Plugin - base class of App::derived::Plugin

=head1 SYNOPSIS

  package App::derived::Plugin::Dumper;
  
  use strict;
  use warnings;
  use parent qw/App::derived::Plugin/;
  use Data::Dumper;
  use Class::Accessor::Lite (
      ro => [qw/interval/]
  );
  
  sub int {
      my $self = shift;
      $self->interval(10) unless $self->interval;

      $self->add_worker(
          'dumper',
          sub {
              while (1) {
                  sleep $self->interval;
                  my @keys = $self->service_keys();
                  for my $key ( @keys ) {
                      my $ref = $self->service_stats->{$key}
                      print Dumper([$key,$ref]);
                  }
              }
          }  
      );
  }


=head1 DESCRIPTION

This module is base class to make App::derived Plugin.

=head1 METHODS

=over 4

=item new

=item json(): Object

Utility method. returns JSON.pm object

=item service_stats($key): HashRef

Retrieve service status

=item exists_service($key): Bool

Checking existence of service named $key

=item service_keys(): Array

get all registered service keys

=item add_worker($tag:String, $code:SubRef)

Registering a worker

=item init

Required to implementing this method in your Plugin

=back

=head1 SEE ALSO

<App::derived::Plugin::Memcached>

=head1 AUTHOR

Masahiro Nagano E<lt>kazeburo@gmail.comE<gt>

=head1 LICENSE

Copyright (C) Masahiro Nagano

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut




lib/App/derived/Plugin/Dumper.pm
package App::derived::Plugin::Dumper;

use strict;
use warnings;
use parent qw/App::derived::Plugin/;
use Class::Accessor::Lite (
    rw => [qw/interval/],
);

sub init {
    my $self = shift;
    $self->interval(10) unless $self->interval;
    $self->add_worker(
        'dumper', sub {
            sleep $self->interval;
            my @keys = $self->service_keys;
            for my $key ( @keys ) {
                my $ref = $self->service_stats($key);
                print $self->json->encode([$key,$ref]), "\n"; 
            }
        }
    );
}

1;

__END__

=encoding utf8

=head1 NAME

App::derived::Plugin::Dumper - Disply serialized data

=head1 SYNOPSIS

  $ derived -MDumper,interval=10 CmdsFile

=head1 DESCRIPTION

This plugin displys serialized data

=head1 ARGUMENTS

=over 4

=item interval:Int

Interval seconds to post

=back
  
=head1 SEE ALSO

<drived>, <App::derived::Plugin> for writing plugins

=head1 AUTHOR

Masahiro Nagano E<lt>kazeburo@gmail.comE<gt>

=head1 LICENSE

Copyright (C) Masahiro Nagano

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut






lib/App/derived/Plugin/Memcached.pm
package App::derived::Plugin::Memcached;

use strict;
use warnings;
use parent qw/App::derived::Plugin/;
use Class::Accessor::Lite (
    ro => [qw/host port timeout/],
);
use IO::Socket::INET;
use POSIX qw(EINTR EAGAIN EWOULDBLOCK :sys_wait_h);
use Socket qw(IPPROTO_TCP TCP_NODELAY);
use Log::Minimal;

our $MAX_REQUEST_SIZE = 131072;
our $CRLF      = "\x0d\x0a";
our $DELIMITER = "\x20";

sub init {
    my $self = shift;

    $self->{host} = '0' unless $self->{host};
    $self->{port} = 12306 unless $self->{port};
    $self->{timeout} = 10 unless $self->{timeout};
    
    my $localaddr = $self->{host} .':'. $self->{port};
    my $sock = IO::Socket::INET->new(
        Listen    => SOMAXCONN,
        LocalAddr => $localaddr,
        Proto     => 'tcp',
        (($^O eq 'MSWin32') ? () : (ReuseAddr => 1)),
    ) or die "failed to listen to port $localaddr: $!";

    $self->add_worker(
        'memcached_server_'.$localaddr,
        sub {
            $0 = "$0 memcached_server";
            $self->server($sock);            
        }
    );
}

sub server {
    my $self = shift;
    my $sock = shift;

    local $SIG{CHLD} = sub {
        1 until (-1 == waitpid(-1, WNOHANG));
    };

    while(1) {
        local $SIG{PIPE} = 'IGNORE';
        if ( my $conn = $sock->accept ) {
            debugf("[server] new connection from %s:%s", $conn->peerhost, $conn->peerport);
            $conn->blocking(0)
                or die "failed to set socket to nonblocking mode:$!";
            $conn->setsockopt(IPPROTO_TCP, TCP_NODELAY, 1)
                or die "setsockopt(TCP_NODELAY) failed:$!";
            my $pid = fork();
            die "cannot fork: $!" unless defined $pid;
            if ( $pid == 0 ) {
                $self->handle_connection($conn);
                debugf("[server] close connection: %s:%s", $conn->peerhost, $conn->peerport);
                $conn->close;
                exit;
            }
            $conn->close;
        }
    }

}

sub handle_connection {
    my ($self, $conn) = @_;
    
    my $buf = '';
    my $req = +{};

    while (1) {
        my $rlen = read_timeout(
            $conn, \$buf, $MAX_REQUEST_SIZE - length($buf), length($buf), $self->{timeout},
        ) or last;
        if ( parse_read_buffer($buf, $req ) ) {
            $buf = '';
            if ( $req->{cmd} eq 'get' ) {
                my @keys = split /\x20+/, $req->{keys};
                my $result;
                debugf("[server] request get => %s from %s:%s", $req->{keys}, $conn->peerhost, $conn->peerport);
                for my $key ( @keys ) {
                    my $mode = '';
                    if ( $key =~ m!:full$! ) {
                        $mode = 'full';
                        $key =~ s!:full$!!;
                    }
                    elsif ( $key =~ m!:latest$! ) {
                        $mode = 'latest';
                        $key =~ s!:latest$!!;                        
                    }
                    if ( $self->exists_service($key) ) {
                        my $ref = $self->service_stats($key);
                        if ( $mode eq 'full' ) {
                            my $val = $self->json->encode($ref);
                            $result .= join $DELIMITER, "VALUE", $key, 0, length($val);
                            $result .= $CRLF . $val . $CRLF;
                        }
                        elsif ( $mode eq 'latest' ) {
                            if ( defined $ref->{latest} ) {
                                my $val = $ref->{latest};
                                $result .= join $DELIMITER, "VALUE", $key, 0, length($val);
                                $result .= $CRLF . $val . $CRLF;
                            }
                        }
                        else {
                            if ( defined $ref->{persec} ) {
                                my $val = $ref->{persec};
                                $result .= join $DELIMITER, "VALUE", $key, 0, length($val);
                                $result .= $CRLF . $val . $CRLF;
                            }
                        }
                    }
                }
                $result .= "END" . $CRLF;
                write_all( $conn, $result, $self->{timeout} );
            }
            elsif ( $req->{cmd} eq 'version' ) {
                write_all( $conn, "VERSION $App::derived::VERSION$CRLF", $self->{timeout} );
            }
            elsif ( $req->{cmd} eq 'quit' ) {
                #do nothing
                last;
            }
            else {
                write_all( $conn, "ERROR".$CRLF, $self->{timeout} );
            }
        }
    }
    return;
}

sub parse_read_buffer {
    my ($buf, $ret) = @_;
    if ( $buf =~ /$CRLF$/o ) {
        my ($req_line) = split /$CRLF/, $buf;
        ($ret->{cmd}, $ret->{keys}) = split /$DELIMITER/o, $req_line, 2;
        $ret->{keys} ||= '';
        return 1;
    }
    return;
}

# returns (positive) number of bytes read, or undef if the socket is to be closed
sub read_timeout {
    my ($sock, $buf, $len, $off, $timeout) = @_;
    do_io(undef, $sock, $buf, $len, $off, $timeout);
}

# returns (positive) number of bytes written, or undef if the socket is to be closed
sub write_timeout {
    my ($sock, $buf, $len, $off, $timeout) = @_;
    do_io(1, $sock, $buf, $len, $off, $timeout);
}

# writes all data in buf and returns number of bytes written or undef if failed
sub write_all {
    my ($sock, $buf, $timeout) = @_;
    my $off = 0;
    while (my $len = length($buf) - $off) {
        my $ret = write_timeout($sock, $buf, $len, $off, $timeout)
            or return;
        $off += $ret;
    }
    return length $buf;
}

# returns value returned by $cb, or undef on timeout or network error
sub do_io {
    my ($is_write, $sock, $buf, $len, $off, $timeout) = @_;
    my $ret;
 DO_READWRITE:
    # try to do the IO
    if ($is_write) {
        $ret = syswrite $sock, $buf, $len, $off
            and return $ret;
    } else {
        $ret = sysread $sock, $$buf, $len, $off
            and return $ret;
    }
    unless ((! defined($ret)
                 && ($! == EINTR || $! == EAGAIN || $! == EWOULDBLOCK))) {
        return;
    }
    # wait for data
 DO_SELECT:
    while (1) {
        my ($rfd, $wfd);
        my $efd = '';
        vec($efd, fileno($sock), 1) = 1;
        if ($is_write) {
            ($rfd, $wfd) = ('', $efd);
        } else {
            ($rfd, $wfd) = ($efd, '');
        }
        my $start_at = time;
        my $nfound = select($rfd, $wfd, $efd, $timeout);
        $timeout -= (time - $start_at);
        last if $nfound;
        return if $timeout <= 0;
    }
    goto DO_READWRITE;
}

1;


__END__

=encoding utf8

=head1 NAME

App::derived::Plugin::Memcached - memcached-protocol server for derived

=head1 SYNOPSIS

  $ derived -MMemcahced,port=12306 CmdsFile

=head1 DESCRIPTION

This plugin has a memcached-protocol server. You can get some status from 
any memcached client.

=head1 ARGUMENTS

=over 4

=item port:Int

Port number to bind

=item host:String

Host name or Address to bind

=item timeout:Int

Timeout seconds to read request. default 10.

=back

=head1 CLIENT SAMPLE

You can access to data by any memcached client.

  use Cache::Memcached::Fast;

  my $memcached = Cache::Memcached::Fast->new({
    servers => [qw/localhost:12306/],
  });

  say $memcached->get('slowqueris'); # only per seconds value.
  say $memcached->get('slowqueris:latest'); # only latest value.
  say $memcached->get('slowqueris:full'); #JSON formated data include raw values
  
=head1 SEE ALSO

<drived>, <App::derived::Plugin> for writing plugins

=head1 AUTHOR

Masahiro Nagano E<lt>kazeburo@gmail.comE<gt>

=head1 LICENSE

Copyright (C) Masahiro Nagano

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut



lib/App/derived/Plugin/GrowthForecast.pm
package App::derived::Plugin::GrowthForecast;

use strict;
use warnings;
use parent qw/App::derived::Plugin/;
use Class::Accessor::Lite (
    rw => [qw/api_url service section type mode interval timeout match_key/],
);
use LWP::UserAgent;
use Log::Minimal;

sub init {
    my $self = shift;
    die '[GrowthForecast] api_url is not defined' unless $self->api_url;
    die '[GrowthForecast] service is not defined' unless $self->service;
    die '[GrowthForecast] section is not defined' unless $self->service;

    $self->interval(10) unless $self->interval;
    $self->timeout(10) unless $self->timeout;
    $self->mode('gauge') unless $self->mode;
    $self->type('latest') unless $self->type;
    $self->match_key('^.+$') unless $self->match_key;

    $self->add_worker(
        'growthforeacst',
        $self->gf_post
    );
}

sub gf_post {
    my $self = shift;
    my $ua = LWP::UserAgent->new(
        agent => 'App::derived::Plugin::GrowthForecast',
        timeout => $self->timeout
    );

    sub {
        while (1) {
            sleep $self->interval;
            my @keys = $self->service_keys;
            my $match_key = $self->match_key;
            for my $key ( @keys ) {
                next if $key !~ m!$match_key!;
                my $number = $self->service_stats($key)->{$self->type};
                next if $number eq '0E0';
                debugf('[GrowthForecast] post %s {number:%s,mode:%s}', $key, $number, $self->mode);
                my $res = $ua->post(
                    $self->api_url . $self->service . '/' . $self->section . '/' . $key, {
                        number => int($number),
                        mode => $self->mode
                    }
                );
                warnf('[GrowthForecast] failed post to %s {number:%s,mode:%s}: %s',
                    $self->api_url . $self->service . '/' . $self->section . '/' . $key,
                    $number, $self->mode, $res->status_line) unless $res->is_success;
                
            }
        }
    }
}

1;

__END__

=encoding utf8

=head1 NAME

App::derived::Plugin::GrowthForecast - post data to GrowthForecast

=head1 SYNOPSIS

  $ derived -MGrowthForecast,api_url=http://host/api,service=service1,section=section1,mode=gauge CmdsFile

=head1 DESCRIPTION

This plugin post data to GrowthForecast 

=head1 ARGUMENTS

=over 4

=item api_url:String

Endpoint of GrowthForecast API

=item service:String

Service name

=item section:String

Section name

=item mode:String

mode option available with gauge(default), count, modified, just same as mode of GrowthForecast POST parameter.

=item type:String

Data type available with "latest"(default) and "persec".

=item interval:Int

Interval seconds to post

=item match_key:Regexp

Post match keys only. default '.+' (all keys)

=back
  
=head1 SEE ALSO

<drived>, <App::derived::Plugin> for writing plugins

=head1 AUTHOR

Masahiro Nagano E<lt>kazeburo@gmail.comE<gt>

=head1 LICENSE

Copyright (C) Masahiro Nagano

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut




lib/App/derived.pm
package App::derived;

use strict;
use warnings;
use 5.008005;
use File::Temp qw/tempfile/;
use File::Copy;
use Proclet;
use JSON ();
use Log::Minimal;

our $VERSION = '0.09';

my $_JSON = JSON->new()
    ->utf8(1)
    ->shrink(1)
    ->space_before(0)
    ->space_after(0)
    ->indent(0);

sub new {
    my $class = shift;
    my %opt = ref $_[0] ? %{$_[0]} : @_;
    my %args = (
        proclet  => Proclet->new(enable_log_worker => ($ENV{LM_DEBUG} ? 1 : 0)),
        interval => 10,
        host     => 0,
        port     => 12306,
        timeout  => 10,
        services => {},
        %opt
    );
    bless \%args, $class;
}

sub add_service {
    my $self = shift;
    my ($key, $cmd) = @_;
    my ($tmpfh,$tmpfile) = tempfile(UNLINK=>0, EXLOCK=>0);
    print $tmpfh $_JSON->encode({
        status=>"INIT",
        persec => '0E0',
        latest => '0E0',
    });
    close $tmpfh;
    $self->{services}->{$key} = {
        cmd => ['bash', '-c', $cmd],
        file => $tmpfile,
        prev => undef,
    };
    infof("register service: %s", $key);
    $self->{proclet}->service(
        code => sub {
            $0 = "$0 worker $key";
            $self->worker($key);
            exit;
        },
        tag => $key.'_worker',
    );
}

sub add_plugin {
    my $self = shift;
    my ( $plugin, $args) = @_;

    my %args = (
        %$args,
        _services => $self->{services},
        _proclet => $self->{proclet},
    );    
    infof("register plugin: %s", $plugin);
    my $instance = $plugin->new(\%args);
    $instance->init();
}

sub run {
    my $self = shift;
    $self->{proclet}->run;
}

sub DESTROY {
    my $self = shift;
    for my $key ( keys %{$self->{services}} ) {
        unlink $self->{services}->{$key}->{file};
    }
}

sub worker {
    my ($self, $service_key) = @_;
    srand();
    my $service = $self->{services}->{$service_key};
    my $n = time;
    $n = $n - ( $n % $self->{interval}) + $self->{interval} + int(rand($self->{interval}));; #next + random
    my $stop = 1;
    local $SIG{TERM} = sub { $stop = 0 };

    while ( $stop ) {
        my $current = time();
        while ( $n < $current ) {
            $n = $n + $self->{interval};
        }
        while ( $stop ) {
            last if time() >= $n;
            select undef, undef, undef, 0.1 ## no critic;
        }
        $n = $n + $self->{interval};
        local $Log::Minimal::AUTODUMP = 1;
        debugf("exec command for %s => %s", $service_key, $service);
        my ($result, $exit_code) = cap_cmd($service->{cmd});
        debugf("command [%s]: exit_code:%s result:%s", $service_key, $exit_code, $result);
        if ( ! defined $result ) {
            atomic_write($service->{file}, {
                status => "ERROR",
                persec => undef,
                latest => undef,
                raw => undef,
                exit_code => $exit_code,
                last_update => time,
            });
            next;
        }
    
        my $orig = $result;
        $result =~ s!^[^0-9]+!!;
        {
            no warnings;
            $result = int($result);
        }
        if ( ! defined $service->{prev} ) {
            $service->{prev} = $result;
            atomic_write( $service->{file}, {
                status => "CALCURATE",
                persec => "E0E",
                latest => $result,
                raw => $orig,
                exit_code => $exit_code,
                last_update => time,
            });
            next;
        }
        my $derive = ($result - $service->{prev}) / $self->{interval};
        atomic_write( $service->{file}, {
            status => "OK",
            persec => $derive,
            latest => $result,
            raw => $orig,
            exit_code => $exit_code,
            last_update => time,
        });
        $service->{prev} = $result;
    }
}

sub cap_cmd {
    my ($cmdref) = @_;
    pipe my $logrh, my $logwh
        or die "Died: failed to create pipe:$!";
    my $pid = fork;
    if ( ! defined $pid ) {
        die "Died: fork failed: $!";
    } 

    elsif ( $pid == 0 ) {
        #child
        close $logrh;
        open STDOUT, '>&', $logwh
            or die "Died: failed to redirect STDOUT";
        close $logwh;
        exec @$cmdref;
        die "Died: exec failed: $!";
    }
    close $logwh;
    my $result;
    while(<$logrh>){
        chomp;chomp;
        $result .= $_;
    }
    close $logrh;
    while (wait == -1) {}
    my $exit_code = $?;
    $exit_code = $exit_code >> 8;
    if ( $exit_code != 0 ) {
        warnf("Error: command exited with code: $exit_code");
    }
    return ($result, $exit_code);
}

sub atomic_write {
    my ($writefile, $body) = @_;
    my ($tmpfh,$tmpfile) = tempfile(UNLINK=>0);
    print $tmpfh $_JSON->encode($body);
    close($tmpfh);
    move( $tmpfile, $writefile);
}


1;
__END__

=encoding utf8

=head1 NAME

App::derived - run command periodically and calculate rate and check from network

=head1 SYNOPSIS

  $ cat CmdsFile
  slowqueries: mysql -NB -e 'show global status like "Slow_queries%"'
  $ derived -MMemcahced,port=11211,host=127.0.0.1 CmdsFile

  $ telnet localhost port
  get slowqueris
  VALUE slowqueris 0 3
  0.2  # slow queries/sec

=head1 DESCRIPTION

derived runs commands periodically and capture integer value. And calculate per-second rate. 
You can retrieve these values from integrated memcached-protocol server or pluggable workers.

You can monitoring the variation of metrics through this daemon.

See detail for perldoc "derived"

=head1 AUTHOR

Masahiro Nagano E<lt>kazeburo@gmail.comE<gt>

=head1 SEE ALSO

<derived>

=head1 LICENSE

Copyright (C) Masahiro Nagano

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut
